src/psyclone/tests/domain/nemo/transformations/nemo_arrayrange2loop_trans_test.py

```diff
@@ -331,20 +332,115 @@ def test_apply_structure_of_arrays():
+    trans.apply(array_ref.member.member.children[0])
+    result = fortran_writer(schedule)
+    assert (
+        "  do jk = 1, jpk, 1\n"
+        "    do jj = 1, jpj, 1\n"
+        "      do ji = 1, jpi, 1\n"
+        "        mystruct%field2%field(ji,jj,jk) = 0.0d0\n"
+        "      enddo\n"
+        "    enddo\n"
+        "  enddo\n" in result)
 
-    # Case 3: Nested SoA currently causes an InternalError
+    # Case 3: AoSoA in the LHS
     array_ref = assignment3.lhs
-    with pytest.raises(InternalError) as info:
+    trans.apply(array_ref.member.children[3])
+    trans.apply(array_ref.member.children[2])
+    trans.apply(array_ref.member.children[1])
+    result = fortran_writer(schedule)
+    assert (
+        "  do jk = 1, jpk, 1\n"
+        "    do jj = 1, jpj, 1\n"
+        "      do ji = 1, jpi, 1\n"
+        "        mystruct%field3(ji,jj,jk)%field4 = 0.0d0\n"
+        "      enddo\n"
+        "    enddo\n"
+        "  enddo\n" in result)
+
+    # Case 4: AoSoA in the LHS and SoA in the RHS
+    array_ref = assignment4.lhs
+    trans.apply(array_ref.member.children[3])
+    trans.apply(array_ref.member.children[2])
+    trans.apply(array_ref.member.children[1])
+    result = fortran_writer(schedule)
+    assert (
+        "  do jk = 1, jpk, 1\n"
+        "    do jj = 1, jpj, 1\n"
+        "      do ji = 1, jpi, 1\n"
+        "        mystruct%field3(ji,jj,jk)%field4 = "
+        "mystruct%field2%field(ji,jj,jk)\n"
+        "      enddo\n"
+        "    enddo\n"
+        "  enddo\n" in result)
+
+
+def test_apply_structure_of_arrays_multiple_arrays(fortran_writer):
+    '''Check that the apply method works when the assignment expression
+    contains structures of arrays with multiple array accessors.
+
+    '''
+    _, invoke_info = get_invoke("implicit_do_structures.f90", api=API, idx=0)
+    schedule = invoke_info.schedule
+    assignment5 = schedule[4]
+    trans = NemoArrayRange2LoopTrans()
+
+    # Case 5: 2 array accessors in LHS but only one has ranges
+    array_ref = assignment5.lhs
+    trans.apply(array_ref.member.member.children[2])
+    trans.apply(array_ref.member.member.children[1])
+    trans.apply(array_ref.member.member.children[0])
+    result = fortran_writer(schedule)
+    assert (
+        "  do jk = 1, jpk, 1\n"
+        "    do jj = 1, jpj, 1\n"
+        "      do ji = 1, jpi, 1\n"
+        "        mystruct%field2(4,3)%field(ji,jj,jk) = "
+        "mystruct%field2(5,8)%field(ji,jj,jk)\n"
+        "      enddo\n"
+        "    enddo\n"
+        "  enddo\n" in result)
+
+
+def test_apply_nested_structure_of_arrays():
+    '''Check that nested structure_of_arrays are not supported. '''
+    _, invoke_info = get_invoke("implicit_do_structures.f90", api=API, idx=0)
+    schedule = invoke_info.schedule
+    assignment6 = schedule[5]
+    assignment7 = schedule[6]
+    assignment8 = schedule[7]
+    trans = NemoArrayRange2LoopTrans()
+
+    # Case 6: 2 array accessors in LHS and both have ranges
```

Comment:
Again, this isn't valid in Fortran.

Reply:
I do agree that this is invalid Fortran and some frontend (fparser or fparserreader) should prevent it, so I deleted the Fortran code. I am not sure if this can never happen if somebody misconstructs PSyIR, so I misconstructed a tree for this test. I think this should stay at least until the tree misconstruction fails by itself.

Reply:
Agreed. Thanks.